---
title: "A tour of package miRNARefine"
author: "Feiyang Sun"
date: "`r {Sys.setlocale('LC_TIME', 'C'); format(Sys.time(), '%d %b %Y')}`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{A tour of package miRNARefine}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

`miRNARefine` is an R package which aims to identify and correct potential data 
irregularities such as outliers, batch effects, and instability in miRNA 
expression across samples, and to provide diagnostic visualizations that help 
researchers assess data reliability before downstream analyses.

To download **miRNARefine**, use the following commands:

```r
require("devtools")
devtools::install_github("feiyangsun246/miRNARefine", build_vignettes = TRUE)
library(miRNARefine)
```

To list all sample functions available in the package:
``` r
ls("package:miRNARefine")
```

To list all sample datasets available in the package:
``` r
data(package = "miRNARefine")
```
There are 2 sample datasets in the package named *miRNASeq1* and *miRNASeq2*.

*miRNASeq1* is a small subset of the BRCA.miRNASeq dataset from the 
`RTCGA.miRNASeq` package. It contains expression information for 10 samples and 
18 miRNAs.

*miRNASeq2* is a small toy miRNA dataset including missing values and outliers. 
It is designed for testing preprocessing and filtering functions. Each row 
represents a sample, and each column represents a miRNA feature.

<br>

## Automatic filter of miRNAs

### description
Function `adaptiveFiltering` automatically filters miRNAs based on expression, 
variance, or missing values. Adjusts thresholds according to the dataset’s 
characteristics to retain informative features while removing noise.

### usage
``` r
# Using miRNASeq data available with package
data(miRNASeq1)
filtered_data <- adaptiveFiltering(miRNAdata = miRNASeq1,
                                   min_expression = NULL,
                                   min_variance = NULL,
                                   max_na = 0.2,
                                   report_summary = TRUE)
filtered_data
```

## Outlier detect and missing value handling

### description
Function `detectOutliersPCA` identifies anomalous samples using principal 
component analysis (Jolliffe, 2002). Flags samples that deviate strongly from 
the majority, helping to improve overall data quality.

Function `missingValueHandling` detects and imputes missing values in the miRNA 
dataset. Supports adaptive methods like mean, median, or KNN imputation 
(Campesato, 2023), ensuring completeness for downstream analysis.

### usage
``` r
# Using miRNASeq data available with package
data(miRNASeq1)
data(miRNASeq2)

result <- detectOutliersPCA(miRNAdata = miRNASeq1,
                            n_components = 2,
                            row_threshold = 0.99,
                            z_threshold = 3,
                            whether_scale = TRUE,
                            report_summary = TRUE)
result
which(result$sample_outliers)

filled <- missingValueHandling(miRNAdata = miRNASeq2,
                               method = "median",
                               k = 5,
                               report_summary = TRUE)
head(filled)
```
## Compare normalization methods

### description
Function `compareNormalization` applies multiple normalization methods (e.g., 
log2, z-score, quantile) and compares their effects. Helps select the most 
appropriate normalization strategy for consistent and comparable expression 
data.

### usage
``` r
# Using miRNASeq data available with package
data(miRNASeq1)
result <- compareNormalization(miRNAdata = miRNASeq1,
                               choose_best = TRUE)
result$normalized$log2[1:5, 1:5]
result$best_method
```

## Batch effect detect

### description
Function `detectBatch` detects and optionally corrects batch effects in the 
dataset. Ensures that technical or experimental batch differences do not 
confound downstream analyses.

### usage
``` r
# Using miRNASeq data available with package
data(miRNASeq2)
filled2 <- missingValueHandling(miRNAdata = miRNASeq2,
                                method = "median",
                                report_summary = FALSE)
result <- detectBatch(miRNAdata = filled2, correct = TRUE,
                      report_summary = FALSE)
result$corrected_data
result$batch_effects
```

## miRNA stability calculation

### description
Function `miRNAStability` calculates feature-level stability metrics such as 
coefficient of variation (CV) or median absolute deviation (MAD). Identifies 
the most reliable miRNAs for robust downstream analyses (Hima Bindu et al., 
2019).

### usage
``` r
# Using miRNASeq1 available with package
data(miRNASeq1)
result <- miRNAStability(miRNAdata = miRNASeq1, report_summary = FALSE)

result$most_stable
result$least_stable
```

## Generate stability distribution plot

### description
Function `plotStabilityDistribution` visualizes the distribution of miRNA 
stability metrics across all features. Highlights highly stable or highly 
variable miRNAs, helping users quickly assess feature quality.

### usage
``` r
# Using miRNASeq1 available with package
data(miRNASeq1)
result1 <- miRNAStability(miRNAdata = miRNASeq1, report_summary = FALSE)
plot1 <- plotStabilityDistribution(stability_results = result1)

# print(plot1)
```

<br>

## Package References

Bolstad B (2025). `preprocessCore`: A collection of pre-processing functions. R package version 1.72.0. <https://bioconductor.org/packages/preprocessCore>

Castelluzzo M, Detassis S, Denti M, Ricci L (2023). `MiRNAQCD`: Micro-RNA Quality Control and Diagnosis. R package version 1.1.3.<https://CRAN.R-project.org/package=MiRNAQCD>

Chodor W. (2025). `RTCGA.miRNASeq`: miRNASeq datasets from The Cancer Genome Atlas Project. R package version 1.36.0. <https://bioconductor.org/packages/RTCGA.miRNASeq>

CRAN Project (2023). `factoextra`: Extract and visualize the results of multivariate data analyses. <https://CRAN.R-project.org/package=factoextra>

Hastie T, Tibshirani R, Narasimhan B, Chu G (2025). `impute`: Imputation for microarray data. doi:10.18129/B9.bioc.impute. R package version 1.85.0, <https://bioconductor.org/packages/impute>

Leek JT, Johnson WE, Parker HS, Fertig EJ, Jaffe AE, Zhang Y, Storey JD, Torres LC (2025). `sva`: Surrogate Variable Analysis. doi:10.18129/B9.bioc.sva, R package version 3.58.0, <https://bioconductor.org/packages/sva>

R Core Team. (2025). Package `stats`. *R: A Language and Environment for Statistical Computing.* R Foundation for Statistical Computing, Vienna, Austria. <https://www.R-project.org/>

Wickham, H. (2016). `ggplot2`: Elegant graphics for data analysis. Springer-Verlag New York. <https://ggplot2.tidyverse.org>

## Other References

Bolstad, B. M., Irizarry, R. A., Astrand, M., & Speed, T. P. (2003). A comparison of normalization methods for high density oligonucleotide array data based on variance and bias. Bioinformatics, 19(2), 185–193.

Hastie, T., Tibshirani, R., & Friedman, J. (2009). The Elements of Statistical Learning, 2nd Edition. Springer.

Hima Bindu, A., Suresh, P., Reddy, R., et al. (2019). Systematic evaluation of miRNA stability across biological replicates. Frontiers in Molecular Biosciences, 6, 153.

Johnson, W. E., Li, C., & Rabinovic, A. (2007). Adjusting batch effects in microarray expression data using empirical Bayes methods. Biostatistics, 8(1), 118-127.

Jolliffe, I. T. (2002). Principal Component Analysis. Springer.

Leek, J. T., et al. (2010). Tackling the widespread and critical impact of batch effects in high-throughput data. Nature Reviews Genetics, 11(10), 733-739.

Li, R., et al. (2016). Evaluation of microRNA expression stability for normalization in quantitative PCR assays. Biochemical and Biophysical Research Communications}, 471(4), 567-574.

OpenAI. (2025). ChatGPT (Nov 3 version) [Large language model]. <https://chat.openai.com/>

