---
title: "A tour of package miRNARefine"
author: "Feiyang Sun"
date: "`r {Sys.setlocale('LC_TIME', 'C'); format(Sys.time(), '%d %b %Y')}`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{A tour of package miRNARefine}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

`miRNARefine` is an R package which aims to identify and correct potential data 
irregularities such as outliers, batch effects, and instability in miRNA 
expression across samples, and to provide diagnostic visualizations that help 
researchers assess data reliability before downstream analyses.

To download **miRNARefine**, use the following commands:

```r
require("devtools")
devtools::install_github("feiyangsun246/miRNARefine", build_vignettes = TRUE)
library(miRNARefine)
```

To list all sample functions available in the package:
``` r
ls("package:miRNARefine")
```

To list all sample datasets available in the package:
``` r
data(package = "miRNARefine")
```
There are 2 sample datasets in the package named *miRNASeq1* and *miRNASeq2*.

*miRNASeq1* is a small subset of the BRCA.miRNASeq dataset from the 
`RTCGA.miRNASeq` package. It contains expression information for 10 samples and 
18 miRNAs.

*miRNASeq2* is a small toy miRNA dataset including missing values and outliers. 
It is designed for testing preprocessing and filtering functions. Each row 
represents a sample, and each column represents a miRNA feature.

<br>

## Automatically filters miRNAs based on expression, variance, or missing Values

### description
Function `adaptiveFiltering` automatically filters miRNAs based on expression, 
variance, or missing values. Adjusts thresholds according to the dataset’s 
characteristics to retain informative features while removing noise.

### usage


## Package References

Bolstad B (2025). `preprocessCore`: A collection of pre-processing functions. R package version 1.72.0. <https://bioconductor.org/packages/preprocessCore>

Castelluzzo M, Detassis S, Denti M, Ricci L (2023). `MiRNAQCD`: Micro-RNA Quality Control and Diagnosis. R package version 1.1.3.<https://CRAN.R-project.org/package=MiRNAQCD>

Chodor, W. (2025). `RTCGA.miRNASeq`: miRNASeq datasets from The Cancer Genome Atlas Project. R package version 1.36.0. <https://bioconductor.org/packages/RTCGA.miRNASeq>

CRAN Project (2023). `factoextra`: Extract and visualize the results of multivariate data analyses. <https://CRAN.R-project.org/package=factoextra>

## Other References

Bolstad, B. M., Irizarry, R. A., Astrand, M., & Speed, T. P. (2003). A comparison of normalization methods for high density oligonucleotide array data based on variance and bias. Bioinformatics, 19(2), 185–193.

Hastie, T., Tibshirani, R., & Friedman, J. (2009). The Elements of Statistical Learning, 2nd Edition. Springer.

Hima Bindu, A., Suresh, P., Reddy, R., et al. (2019). Systematic evaluation of miRNA stability across biological replicates. Frontiers in Molecular Biosciences, 6, 153.

Johnson, W. E., Li, C., & Rabinovic, A. (2007). Adjusting batch effects in microarray expression data using empirical Bayes methods. Biostatistics, 8(1), 118-127.

Jolliffe, I. T. (2002). Principal Component Analysis. Springer.

Leek, J. T., et al. (2010). Tackling the widespread and critical impact of batch effects in high-throughput data. Nature Reviews Genetics, 11(10), 733-739.

Li, R., et al. (2016). Evaluation of microRNA expression stability for normalization in quantitative PCR assays. Biochemical and Biophysical Research Communications}, 471(4), 567-574.
